### PR TITLE
Higher order functions

### PR DESCRIPTION
**Pull request by JosVerburg**
Implemented the higher order functions:
- fn:for-each
- fn:fold-left
- fn:fold-right

### CODE REVIEW

File: src/expressions/functions/builtInFunctions.sequences.js [hunk 1]
```diff
@@ -469,6 +474,133 @@ function fnFilter (dynamicContext, executionParameters, staticContext, sequence,
...
+	});
+
+	const outerIterator = sequences.value();
+	let innerIterator;
+	return new Sequence({
+		next: () => {
+			while (true) {
+				if (!innerIterator) {
+					const nextSequence = outerIterator.next();
+					// TODO: Consider handling the ready state
```
Comment (DrRataplan) [inline src/expressions/functions/builtInFunctions.sequences.js]: Yup, please handle it by returning the nextSequence variable if the ready property is set to false.

File: src/expressions/functions/builtInFunctions.sequences.js [hunk 2]
```diff
@@ -469,6 +474,133 @@ function fnFilter (dynamicContext, executionParameters, staticContext, sequence,
...
+	 * @type {FunctionValue}
+	 */
+	const callbackFn = callbackSequence.first();
+	const callbackArgumentTypes = callbackFn.getArgumentTypes();
+	if (callbackArgumentTypes.length !== 2) {
+		throw new Error(`XPTY0004: signature of function passed to fn:fold-right is incompatible.`);
+	}
+
+	return sequence.mapAll(values =>
+		values.reverse().reduce((previous, current) => {
```
Comment (DrRataplan) [inline src/expressions/functions/builtInFunctions.sequences.js]: You could use reduceRight here...

File: src/expressions/functions/builtInFunctions.sequences.js [hunk 3]
```diff
@@ -449,10 +449,15 @@ function fnFilter (dynamicContext, executionParameters, staticContext, sequence,
 	 * @type {FunctionValue}
 	 */
 	const callbackFn = callbackSequence.first();
+	const callbackArgumentTypes = callbackFn.getArgumentTypes();
+	if (callbackArgumentTypes.length !== 1) {
+		throw new Error(`XPTY0004: signature of function passed to fn:filter is incompatible.`);
```
Comment (DrRataplan) [inline src/expressions/functions/builtInFunctions.sequences.js]: Nice, I like this!

File: src/expressions/functions/builtInFunctions.sequences.js [hunk 4]
```diff
@@ -668,8 +800,32 @@ export default {
 			namespaceURI: FUNCTIONS_NAMESPACE_URI,
 			localName: 'filter',
 			argumentTypes: ['item()*', 'function(*)'],
-			returnType: 'item()',
+			returnType: 'item()*',
```
Comment (DrRataplan) [inline src/expressions/functions/builtInFunctions.sequences.js]: Good find!

File: test/runnableTestSets.csv [hunk 5]
```diff
@@ -83,12 +83,12 @@ fn-exists,true
...
 fn-floor,true
-fn-fold-left,false
-fn-fold-right,false
-fn-for-each,false
+fn-fold-left,true
+fn-fold-right,true
+fn-for-each,true
 fn-for-each-pair,false
 fn-format-date,false
-fn-format-dateTime,flase
```
Comment (DrRataplan) [inline test/runnableTestSets.csv]: Another good find!

**Comment by JosVerburg** [comment]
@DrRataplan PR comments have been resolved

**Pull request merged**